contrib_bots/lib/giphy.py

```diff
@@ -0,0 +1,106 @@
+    giphy. It looks for messages starting with "@giphy" or the mention that led
+    to the bot, then we send a gif image that appropriate with keywords. You
+    also can send the keyword via private message with bot.
+    '''
+    def usage(self):
+        return '''
+            This plugin will allow users to insert gif images from giphy that
+            appropriate with keywords user input. Users should preface keywords
+            with "@giphy" or the mention that led to the bot. You also can send
+            the keyword via private message with bot.
+            '''
+
+    def get_url_gif_giphy(self, keyword, api_key):
+        # Return gif giphy url based on keywords given, it will return number
+        # if some problem happen, like failed to fetch gif url, or gif url
+        # not found by giphy.
+        query = {'s': keyword,
+                 'api_key': api_key}
+        try:
+            data = requests.get('http://api.giphy.com/v1/gifs/translate', params=query)
+        except requests.exceptions.ConnectionError:  # Usually triggered by bad connection.
+            return 1
+
+        if not data.status_code == 200:
+            return 1
+
+        search_status = data.json()['meta']['status']
+        if not search_status == 200:
+            return 1
+
+        try:
+            gif_url = data.json()['data']['images']['original']['url']
+        except TypeError:  # Usually triggered by no result in Giphy.
+            return 2
+
+        return gif_url
+
+    def triage_message(self, message, bot_name):
+        # Return True if we want to (possibly) response to this message.
+        original_content = message['content']
+        is_giphy_called = (original_content.startswith('@giphy ') or
+                           original_content.startswith('@**%s** ' % (bot_name)))
+
+        # To prevent infinite loop in private message, bot will detect
+        # if the sender name is the bot name it will return false.
+        if message['type'] == 'private':
+            if bot_name == message['sender_full_name']:
+                return False
+            else:
+                return True
+
+        return is_giphy_called
+
+    def handle_message(self, message, client, state_handler):
+        if message['content'].startswith('@giphy '):
+            keyword = message['content'][len('@giphy '):]
+        elif message['content'].startswith('@**%s** ' % (client.bot_name)):
+            keyword = message['content'][len('@**%s** ' % (client.bot_name)):]
+        elif message['type'] == 'private':
+            keyword = message['content']
+
+        # Each error has specified reply, if `gif_url` returns number then bot will
+        # reply the message with specified message below.
+        error_1 = ('Uh oh, sorry :slightly_frowning_face:, I cannot process your request '
+                   'right now. But, let\'s try again later, maybe I can process your '
+                   'request again!  :grin:')
```

Comment:
I think we can drop the `maybe I can process your request again!` and simply leave `But let's try again later!`